irods/connection.py

```diff
@@ -487,16 +489,16 @@ def _login_pam(self):
             if getattr(self,'DISALLOWING_PAM_PLAINTEXT',True):
                 raise PlainTextPAMPasswordError
 
-        # In general authentication API, a ';' and '=' in the password would be misinterpreted due to those
-        # characters' special meaning in the context string parameter.
-        use_dedicated_pam_api = len(ctx) >= MAX_NAME_LEN or \
-                                {';','='}.intersection(set(new_pam_password))
+        use_dedicated_pam_api = (len(ctx) >= MAX_NAME_LEN) or \
+                                cfg.legacy_auth.pam.force_use_of_dedicated_pam_api
 
         if use_dedicated_pam_api:
+            method = "PamAuthRequest"
```

Comment:
Can you write some more words here in a comment or two about when/why this API is used vs the PluginAuthMessage below?

Is this just an 'age of client' thing?   Help the reader with some history here please.

Reply:
We could name it from the "_AN" strings common to both PRC and iRODS api headers? Thoughts?

Reply:
consistency is always good - helps across clients / libraries.

but i just want to know more about why there are two in the first place?   regardless of their names.

Reply:
> consistency is always good - helps across clients / libraries.
> 
> but i just want to know more about why there are two in the first place? regardless of their names.

I see... yeah that happened before my time.  Would be worth discussing. Looks like someone made an attempt at a generic auth plugin way before 4.3 was a thing.

Reply:
https://github.com/irods/python-irodsclient/commit/145f8747394856af48b02461954d83d981e1e131

Does that commit message suggest we can always just use the more flexible method (in 3.0.0)?

Reply:
Ok, so maybe we could pave over https://github.com/irods/python-irodsclient/commit/145f8747394856af48b02461954d83d981e1e131 and just always use `PamAuthRequest` .  How would we then preserve the work I've done in that commit, in case we find we have to backtrack?

Reply:
Just duplicate the branch to a well-established location and record that in our offlines notes?  That probably won't be safe, will it.

Reply:
Our process always preserves the work, so we do not need to worry about that. The commit will always be there because we use git (short of rewriting history via an unallowable rebase / force-push). git-blame makes tracing back changes trivial. Additionally, the original commit message has an issue number so that we have the full context in GitHub. In fact, I will link it here, as well: #279. Now, we will even see this conversation in the future.

The commit which introduces the proposed change can also refer to the sha https://github.com/irods/python-irodsclient/commit/145f8747394856af48b02461954d83d981e1e131 in its message so that investigators in the future can trace back to the change being superseded as well. This is a little extra, but definitely helpful.

Reply:
Oh, perhaps you were not actually referring to the preservation of https://github.com/irods/python-irodsclient/commit/145f8747394856af48b02461954d83d981e1e131 but rather one of the commits in this PR. My bad.

Reply:
Oh, perhaps you were not actually referring to the preservation of https://github.com/irods/python-irodsclient/commit/145f8747394856af48b02461954d83d981e1e131 but rather one of the commits in this PR. My bad.

Reply:
Of course, we could also just keep the setting `legacy_auth.pam.force_use_of_dedicated_pam_api` around, document it, and have users set it to True if they want just the one, guaranteed flexible algorithm to be used.  It's almost the same thing.  Then in 3.1.0, all becomes more moot because we'll be targeting 4.3 only, which uses neither of these APIs.

Reply:
> Oh, perhaps you were not actually referring to the preservation of [145f874](https://github.com/irods/python-irodsclient/commit/145f8747394856af48b02461954d83d981e1e131) but rather one of the commits in this PR. My bad.

It's not your bad, it's mine.  It was lazy assumptions.  You're right, I'm favoring preserving the work in this PR before squashing, so ignore my reference to 145f8747394856af48b02461954d83d981e1e131 please.

Reply:
> Of course, we could also just keep the setting `legacy_auth.pam.force_use_of_dedicated_pam_api` around, document it, and have users set it to True if they want just the one, guaranteed flexible algorithm to be used. It's almost the same thing. Then in 3.1.0, all becomes more moot because we'll be targeting 4.3 only, which uses neither of these APIs.

I think I agree with this.

Reply:
let's do it.

Reply:
Looks like we have a comment and documentation in the README. Shall we resolve this one?

Reply:
Yes!